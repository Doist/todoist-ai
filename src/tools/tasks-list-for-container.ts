import { GetTasksArgs } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { mapTask } from '../tool-helpers.js'
import { generateTaskNextSteps, previewTasks, summarizeList } from '../utils/response-builders.js'
import { TOOL_NAMES } from '../utils/tool-names.js'

const { TASKS_ADD_MULTIPLE } = TOOL_NAMES

const ArgsSchema = {
    type: z
        .enum(['project', 'section', 'parent'])
        .describe('The type of container to get tasks for.'),
    id: z.string().min(1).describe('The ID of the container to get tasks for.'),
    limit: z
        .number()
        .int()
        .min(1)
        .max(50)
        .default(10)
        .describe('The maximum number of tasks to return.'),
    cursor: z
        .string()
        .optional()
        .describe(
            'The cursor to get the next page of tasks (cursor is obtained from the previous call to this tool, with the same parameters).',
        ),
}

const tasksListForContainer = {
    name: TOOL_NAMES.TASKS_LIST_FOR_CONTAINER,
    description: 'Get tasks for a specific project, section, or parent task (subtasks).',
    parameters: ArgsSchema,
    async execute(args, client) {
        const { type, id, limit, cursor } = args

        const taskParams: GetTasksArgs = {
            limit,
            cursor: cursor ?? null,
        }

        switch (type) {
            case 'project':
                taskParams.projectId = id
                break
            case 'section':
                taskParams.sectionId = id
                break
            case 'parent':
                taskParams.parentId = id
                break
        }

        const { results, nextCursor } = await client.getTasks(taskParams)
        const mappedTasks = results.map(mapTask)

        const textContent = generateTextContent({
            tasks: mappedTasks,
            args,
            nextCursor,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                tasks: mappedTasks,
                nextCursor,
                totalCount: mappedTasks.length,
                hasMore: Boolean(nextCursor),
                appliedFilters: {
                    type: args.type,
                    id: args.id,
                    limit: args.limit,
                    cursor: args.cursor,
                },
                containerInfo: {
                    type: args.type,
                    id: args.id,
                },
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    tasks,
    args,
    nextCursor,
}: {
    tasks: ReturnType<typeof mapTask>[]
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
    nextCursor: string | null
}) {
    // Generate container descriptions without API calls
    let subject: string
    let filterHints: string[] = []

    switch (args.type) {
        case 'project':
            subject = 'Tasks in project'
            filterHints = [`in project ${args.id}`]
            break
        case 'section':
            subject = 'Tasks in section'
            filterHints = [`in section ${args.id}`]
            break
        case 'parent':
            subject = 'Subtasks'
            filterHints = [`subtasks of ${args.id}`]
            break
    }

    // Generate helpful suggestions for empty results
    const zeroReasonHints: string[] = []
    if (tasks.length === 0) {
        if (args.type === 'project') {
            zeroReasonHints.push('Project has no tasks yet')
            zeroReasonHints.push(`Use ${TASKS_ADD_MULTIPLE} to create tasks`)
        } else if (args.type === 'section') {
            zeroReasonHints.push('Section is empty')
            zeroReasonHints.push('Tasks may be in other sections of the project')
        } else {
            zeroReasonHints.push('No subtasks created yet')
            zeroReasonHints.push(`Use ${TASKS_ADD_MULTIPLE} with parentId to add subtasks`)
        }
    }

    // Generate contextual next steps
    const now = new Date()
    const todayStr = now.toISOString().split('T')[0]
    const nextSteps = generateTaskNextSteps('listed', tasks, {
        hasToday: tasks.some((task) => task.dueDate === todayStr),
        hasOverdue: tasks.some((task) => task.dueDate && new Date(task.dueDate) < now),
    })

    return summarizeList({
        subject,
        count: tasks.length,
        limit: args.limit,
        nextCursor: nextCursor ?? undefined,
        filterHints,
        previewLines: previewTasks(tasks),
        zeroReasonHints,
        nextSteps,
    })
}

export { tasksListForContainer }
