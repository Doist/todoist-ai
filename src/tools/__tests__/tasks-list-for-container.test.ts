import type { Task, TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { tasksListForContainer } from '../tasks-list-for-container.js'
import {
    TEST_IDS,
    TODAY,
    createMockApiResponse,
    createMockTask,
    extractTextContent,
} from '../test-helpers.js'

// Mock the Todoist API
const mockTodoistApi = {
    getTasks: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

describe('tasks-list-for-container tool', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('container types', () => {
        it.each([
            {
                name: 'project',
                type: 'project' as const,
                id: TEST_IDS.PROJECT_TEST,
                expectedParam: { projectId: TEST_IDS.PROJECT_TEST },
                tasks: [createMockTask({ content: 'Project task' })],
            },
            {
                name: 'section',
                type: 'section' as const,
                id: TEST_IDS.SECTION_1,
                expectedParam: { sectionId: TEST_IDS.SECTION_1 },
                tasks: [createMockTask({ content: 'Section task' })],
            },
            {
                name: 'parent task',
                type: 'parent' as const,
                id: TEST_IDS.TASK_1,
                expectedParam: { parentId: TEST_IDS.TASK_1 },
                tasks: [createMockTask({ content: 'Subtask' })],
            },
        ])('should get tasks for $name', async ({ type, id, expectedParam, tasks }) => {
            mockTodoistApi.getTasks.mockResolvedValue(createMockApiResponse(tasks))

            const result = await tasksListForContainer.execute(
                { type, id, limit: 10 },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                limit: 10,
                cursor: null,
                ...expectedParam,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('getting tasks for a section', () => {
        it('should get tasks for a specific section', async () => {
            const mockTasks: Task[] = [
                createMockTask({
                    id: '8485093751',
                    content: 'Section task 1',
                    description: 'Task in specific section',
                    labels: ['urgent'],
                    childOrder: 1,
                    priority: 3,
                    projectId: '6cfCcrrCFg2xP94Q',
                    sectionId: 'section-123',
                    url: 'https://todoist.com/showTask?id=8485093751',
                    addedAt: '2025-08-13T22:09:59.123456Z',
                    due: {
                        date: '2025-08-16',
                        isRecurring: true,
                        lang: 'en',
                        string: 'every week',
                        timezone: null,
                    },
                }),
            ]

            mockTodoistApi.getTasks.mockResolvedValue({ results: mockTasks, nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'section', id: 'section-123', limit: 15 },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                limit: 15,
                cursor: null,
                sectionId: 'section-123',
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it('should handle empty section', async () => {
            mockTodoistApi.getTasks.mockResolvedValue({ results: [], nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'section', id: 'empty-section-id', limit: 10 },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                limit: 10,
                cursor: null,
                sectionId: 'empty-section-id',
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('getting subtasks for a parent task', () => {
        it('should get subtasks for a specific parent task', async () => {
            const mockSubtasks: Task[] = [
                createMockTask({
                    id: '8485093752',
                    content: 'Subtask 1',
                    description: 'First subtask',
                    childOrder: 1,
                    priority: 1,
                    projectId: '6cfCcrrCFg2xP94Q',
                    parentId: 'parent-task-123',
                    url: 'https://todoist.com/showTask?id=8485093752',
                    addedAt: '2025-08-13T22:10:00.123456Z',
                }),
                createMockTask({
                    id: '8485093753',
                    content: 'Subtask 2',
                    description: 'Second subtask',
                    labels: ['follow-up'],
                    childOrder: 2,
                    priority: 2,
                    projectId: '6cfCcrrCFg2xP94Q',
                    parentId: 'parent-task-123',
                    url: 'https://todoist.com/showTask?id=8485093753',
                    addedAt: '2025-08-13T22:10:01.123456Z',
                    due: {
                        date: '2025-08-18',
                        isRecurring: false,
                        lang: 'en',
                        string: 'Aug 18',
                        timezone: null,
                    },
                }),
            ]

            mockTodoistApi.getTasks.mockResolvedValue({ results: mockSubtasks, nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'parent', id: 'parent-task-123', limit: 10 },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                parentId: 'parent-task-123',
                cursor: null,
                limit: 10,
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })

        it('should handle parent task with no subtasks', async () => {
            mockTodoistApi.getTasks.mockResolvedValue({ results: [], nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'parent', id: 'parent-with-no-subtasks', limit: 10 },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                parentId: 'parent-with-no-subtasks',
                cursor: null,
                limit: 10,
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('pagination and limits', () => {
        it('should handle custom limit and cursor', async () => {
            mockTodoistApi.getTasks.mockResolvedValue({ results: [], nextCursor: 'next-cursor' })

            const result = await tasksListForContainer.execute(
                { type: 'section', id: 'test-section', limit: 25, cursor: 'current-cursor' },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                limit: 25,
                cursor: 'current-cursor',
                sectionId: 'test-section',
            })

            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()
        })
    })

    describe('container type validation', () => {
        it('should handle all container types correctly', async () => {
            mockTodoistApi.getTasks.mockResolvedValue({ results: [], nextCursor: null })

            // Test project container
            await tasksListForContainer.execute(
                { type: 'project', id: 'proj-1', limit: 10 },
                mockTodoistApi,
            )
            expect(mockTodoistApi.getTasks).toHaveBeenLastCalledWith(
                expect.objectContaining({ projectId: 'proj-1' }),
            )

            // Test section container
            await tasksListForContainer.execute(
                { type: 'section', id: 'sect-1', limit: 10 },
                mockTodoistApi,
            )
            expect(mockTodoistApi.getTasks).toHaveBeenLastCalledWith(
                expect.objectContaining({ sectionId: 'sect-1' }),
            )

            // Test parent container
            await tasksListForContainer.execute(
                { type: 'parent', id: 'parent-1', limit: 10 },
                mockTodoistApi,
            )
            expect(mockTodoistApi.getTasks).toHaveBeenLastCalledWith(
                expect.objectContaining({ parentId: 'parent-1' }),
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledTimes(3)
        })
    })

    describe('error handling', () => {
        it('should propagate API errors for project queries', async () => {
            const apiError = new Error('API Error: Project not found')
            mockTodoistApi.getTasks.mockRejectedValue(apiError)

            await expect(
                tasksListForContainer.execute(
                    { type: 'project', id: 'non-existent-project', limit: 10 },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Project not found')
        })

        it('should propagate API errors for section queries', async () => {
            const apiError = new Error('API Error: Section not found')
            mockTodoistApi.getTasks.mockRejectedValue(apiError)

            await expect(
                tasksListForContainer.execute(
                    { type: 'section', id: 'non-existent-section', limit: 10 },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Section not found')
        })

        it('should propagate API errors for parent task queries', async () => {
            const apiError = new Error('API Error: Parent task not found')
            mockTodoistApi.getTasks.mockRejectedValue(apiError)

            await expect(
                tasksListForContainer.execute(
                    { type: 'parent', id: 'non-existent-parent', limit: 10 },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Parent task not found')
        })

        it('should handle permission errors', async () => {
            const permissionError = new Error('API Error: Insufficient permissions')
            mockTodoistApi.getTasks.mockRejectedValue(permissionError)

            await expect(
                tasksListForContainer.execute(
                    { type: 'project', id: 'restricted-project', limit: 10 },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Insufficient permissions')
        })

        it('should handle cursor validation errors', async () => {
            const cursorError = new Error('API Error: Invalid cursor')
            mockTodoistApi.getTasks.mockRejectedValue(cursorError)

            await expect(
                tasksListForContainer.execute(
                    { type: 'project', id: 'test-project', cursor: 'invalid-cursor', limit: 10 },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Invalid cursor')
        })
    })

    describe('zero reason hints for each container type', () => {
        it('should provide project-specific hints for empty projects', async () => {
            mockTodoistApi.getTasks.mockResolvedValue({ results: [], nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'project', id: 'empty-project', limit: 10 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Project has no tasks yet')
            expect(textContent).toContain('Use tasks-add-multiple to create tasks')
        })

        it('should provide section-specific hints for empty sections', async () => {
            mockTodoistApi.getTasks.mockResolvedValue({ results: [], nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'section', id: 'empty-section', limit: 10 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Section is empty')
            expect(textContent).toContain('Tasks may be in other sections of the project')
        })

        it('should provide parent-specific hints for tasks with no subtasks', async () => {
            mockTodoistApi.getTasks.mockResolvedValue({ results: [], nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'parent', id: 'task-no-subtasks', limit: 10 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('No subtasks created yet')
            expect(textContent).toContain('Use tasks-add-multiple with parentId to add subtasks')
        })
    })

    describe('next steps logic', () => {
        it('should suggest appropriate actions when hasOverdue is true', async () => {
            const mockTasks = [
                createMockTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Overdue task in container',
                    due: {
                        date: '2025-08-10',
                        isRecurring: false,
                        lang: 'en',
                        string: 'Aug 10',
                        timezone: null,
                    },
                }),
            ]
            mockTodoistApi.getTasks.mockResolvedValue({ results: mockTasks, nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'project', id: 'project-with-overdue', limit: 10 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(
                'Use tasks-update-multiple to modify priorities or due dates',
            )
            expect(textContent).toContain('Use tasks-complete-multiple to mark finished tasks')
        })

        it('should suggest appropriate actions when hasToday is true', async () => {
            const mockTasks = [
                createMockTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Task due today in container',
                    due: {
                        date: TODAY,
                        isRecurring: false,
                        lang: 'en',
                        string: 'today',
                        timezone: null,
                    },
                }),
            ]
            mockTodoistApi.getTasks.mockResolvedValue({ results: mockTasks, nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'section', id: 'section-with-today', limit: 10 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(
                'Use tasks-update-multiple to modify priorities or due dates',
            )
            expect(textContent).toContain('Use tasks-complete-multiple to mark finished tasks')
        })

        it('should provide standard next steps for regular tasks', async () => {
            const mockTasks = [
                createMockTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Regular future task',
                    due: {
                        date: '2025-08-25',
                        isRecurring: false,
                        lang: 'en',
                        string: 'Aug 25',
                        timezone: null,
                    },
                }),
            ]
            mockTodoistApi.getTasks.mockResolvedValue({ results: mockTasks, nextCursor: null })

            const result = await tasksListForContainer.execute(
                { type: 'parent', id: 'parent-with-regular', limit: 10 },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(
                'Use tasks-update-multiple to modify priorities or due dates',
            )
            expect(textContent).toContain('Use tasks-complete-multiple to mark finished tasks')
        })
    })
})
