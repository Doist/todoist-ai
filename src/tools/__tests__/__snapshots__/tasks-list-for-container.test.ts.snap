// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`tasks-list-for-container tool container types should get tasks for parent task 1`] = `
"Subtasks: 1 (limit 10).
Filter: subtasks of 8485093748.
Preview:
    Subtask • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks"
`;

exports[`tasks-list-for-container tool container types should get tasks for project 1`] = `
"Tasks in project: 1 (limit 10).
Filter: in project 6cfCcrrCFg2xP94Q.
Preview:
    Project task • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks"
`;

exports[`tasks-list-for-container tool container types should get tasks for section 1`] = `
"Tasks in section: 1 (limit 10).
Filter: in section section-123.
Preview:
    Section task • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks"
`;

exports[`tasks-list-for-container tool getting subtasks for a parent task should get subtasks for a specific parent task 1`] = `
"Subtasks: 2 (limit 10).
Filter: subtasks of parent-task-123.
Preview:
    Subtask 1 • P1 • id=8485093752
    Subtask 2 • due 2025-08-18 • P2 • id=8485093753
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`tasks-list-for-container tool getting subtasks for a parent task should handle parent task with no subtasks 1`] = `
"Subtasks: 0 (limit 10).
Filter: subtasks of parent-with-no-subtasks.
No results. No subtasks created yet; Use tasks-add-multiple with parentId to add subtasks."
`;

exports[`tasks-list-for-container tool getting tasks for a section should get tasks for a specific section 1`] = `
"Tasks in section: 1 (limit 15).
Filter: in section section-123.
Preview:
    Section task 1 • due 2025-08-16 • P3 • id=8485093751
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`tasks-list-for-container tool getting tasks for a section should handle empty section 1`] = `
"Tasks in section: 0 (limit 10).
Filter: in section empty-section-id.
No results. Section is empty; Tasks may be in other sections of the project."
`;

exports[`tasks-list-for-container tool next steps logic should provide standard next steps for regular tasks 1`] = `
"Subtasks: 1 (limit 10).
Filter: subtasks of parent-with-regular.
Preview:
    Regular future task • due 2025-08-25 • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks"
`;

exports[`tasks-list-for-container tool next steps logic should suggest appropriate actions when hasOverdue is true 1`] = `
"Tasks in project: 1 (limit 10).
Filter: in project project-with-overdue.
Preview:
    Overdue task in container • due 2025-08-10 • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`tasks-list-for-container tool next steps logic should suggest appropriate actions when hasToday is true 1`] = `
"Tasks in section: 1 (limit 10).
Filter: in section section-with-today.
Preview:
    Task due today in container • due 2025-08-17 • P1 • id=8485093748
Next:
- Use tasks-update-multiple to modify priorities or due dates
- Use tasks-complete-multiple to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`tasks-list-for-container tool pagination and limits should handle custom limit and cursor 1`] = `
"Tasks in section: 0 (limit 25), more available.
Filter: in section test-section.
No results. Section is empty; Tasks may be in other sections of the project.
Next:
- Pass cursor 'next-cursor' to fetch more results."
`;

exports[`tasks-list-for-container tool zero reason hints for each container type should provide parent-specific hints for tasks with no subtasks 1`] = `
"Subtasks: 0 (limit 10).
Filter: subtasks of task-no-subtasks.
No results. No subtasks created yet; Use tasks-add-multiple with parentId to add subtasks."
`;

exports[`tasks-list-for-container tool zero reason hints for each container type should provide project-specific hints for empty projects 1`] = `
"Tasks in project: 0 (limit 10).
Filter: in project empty-project.
No results. Project has no tasks yet; Use tasks-add-multiple to create tasks."
`;

exports[`tasks-list-for-container tool zero reason hints for each container type should provide section-specific hints for empty sections 1`] = `
"Tasks in section: 0 (limit 10).
Filter: in section empty-section.
No results. Section is empty; Tasks may be in other sections of the project."
`;
